<a href="https://colab.research.google.com/github/KonstantinSergeev2/Data-analysis/blob/main/Lab7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Пункт 1 - загружаем данные из файла.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score, precision_score, recall_score

# Загружаем данные из файла "База".
dataF = pd.read_csv('sample_data/База.csv', sep=';', encoding='cp1251')

# Выводим первые 10 строк.
with pd.option_context('display.max_columns', None):
  display(dataF.head(10))

# Узнаем размер данных.
print("Количество строк и столбцов:", dataF.shape)

УИД_Брони   ДатаБрони ВремяБрони ИсточникБрони  \
0  d192173f-fc14-11eb-9512-000c29ad50ac  13.08.2021    1:00:01        ручная   
1  43574a1f-fe8b-11eb-9512-000c29ad50ac  16.08.2021    4:12:46        ручная   
2  0e7b7a81-fe97-11eb-9512-000c29ad50ac  16.08.2021    5:37:12        ручная   
3  c7041428-f90b-11eb-9512-000c29ad50ac  09.08.2021    4:17:35        ручная   
4  60090518-fe8b-11eb-9512-000c29ad50ac  16.08.2021    4:13:35        ручная   
5  9dd1948f-feb7-11eb-9512-000c29ad50ac  16.08.2021    9:30:15            МП   
6  f7eafe38-ffe3-11eb-9512-000c29ad50ac  18.08.2021    9:20:15            МП   
7  8446f94a-00b5-11ec-9512-000c29ad50ac  19.08.2021   10:20:16            МП   
8  005522de-01fb-11ec-9512-000c29ad50ac  21.08.2021    1:10:12            МП   
9  61668cc7-0262-11ec-9512-000c29ad50ac  21.08.2021    1:30:21            МП   

  ВременнаяБронь СледующийСтатус             Город     ВидПомещения   Тип  \
0             Да        Свободна         Ярославль  жилые помещения  2,5к   
1             Да        Свободна         Ярославль  жилые помещения  3,5к   
2             Да        Свободна         Ярославль  жилые помещения  2,5к   
3             Да        Свободна            Тюмень  жилые помещения  2,5к   
4             Да        Свободна         Ярославль  жилые помещения  2,5к   
5            Нет        Свободна  Набережные Челны  жилые помещения  1,5к   
6            Нет        Свободна            Тюмень  жилые помещения  3,5к   
7            Нет        Свободна            Тюмень  жилые помещения    2к   
8            Нет        Свободна            Тюмень         кладовые  клад   
9            Нет        Свободна  Набережные Челны  жилые помещения    2к   

  ПродаваемаяПлощадь  Этаж СтоимостьНаДатуБрони               ТипСтоимости  \
0                 72   3.0              4296100  Стоимость при 100% оплате   
1               79,8   2.0              5279600  Стоимость при 100% оплате   
2               65,4  18.0              4201400  Стоимость при 100% оплате   
3               61,6  12.0              5414500  Стоимость при 100% оплате   
4               65,4  18.0              4201400  Стоимость при 100% оплате   
5                 47   3.0              3494600  Стоимость при 100% оплате   
6               97,4   8.0              7535000  Стоимость при 100% оплате   
7               55,4   4.0              4954700  Стоимость при 100% оплате   
8               3,02   NaN               105700  Стоимость при 100% оплате   
9               52,1   4.0              4030800  Стоимость при 100% оплате   

           ВариантОплаты ВариантОплатыДоп СкидкаНаКвартиру  \
0  Единовременная оплата              NaN              NaN   
1  Единовременная оплата              NaN              NaN   
2  Единовременная оплата              NaN              NaN   
3  Единовременная оплата              NaN              NaN   
4  Единовременная оплата              NaN              NaN   
5  Единовременная оплата          Ипотека              NaN   
6  Единовременная оплата          Ипотека              NaN   
7  Единовременная оплата          Ипотека              NaN   
8  Единовременная оплата              NaN              NaN   
9  Единовременная оплата          Ипотека              NaN   

  ФактическаяСтоимостьПомещения СделкаАН ИнвестиционныйПродукт Привилегия  \
0                       4296100      Нет                   Нет        Нет   
1                       5279600       Да                   Нет        Нет   
2                       4201400      Нет                   Нет        Нет   
3                       5414500      Нет                   Нет        Нет   
4                       4201400      Нет                   Нет        Нет   
5                       3494600      Нет                   Нет        Нет   
6                       7535000       Да                   Нет        Нет   
7                       4954700      Нет                   Нет        Нет   
8                        105700      Нет                   Нет    

Количество строк и столбцов: (5519, 21)


Загружаем данные из файла, обязательно указываем кодировку "cp1251", а не "utf-8" и прочее, так как вывод из данного файла осуществляется в этой кодировке. выводим первые 10 строчек для проверки, с помощью "with pd.option_context('display.max_columns', None):" указываем, что нужно вывести полностью все данные из строк, так как, если этого не указать, то при выводе вместо этажей мы увидим просто троеточие. Также дополнительно узнаём размерность таблицы, чтобы убедиться в том, что все столбцы и строки действительно считались.

In [ ]:
# Пункт 2 - предварительная фильтрация данных.

# Сохраняем исходные данные.
dataF_original = dataF.copy()

# a. Фильтруем только по жилым помещениям.
print("Количество строк до фильтрации по жилым помещениям:", len(dataF))
dataF_new = dataF[dataF['ВидПомещения'] == 'жилые помещения']
print("Количество строк после фильтрации по жилым помещениям:", len(dataF_new))

# Удаляем столбенц ВидПомещения по задаче.
dataF_new = dataF_new.drop(columns=["ВидПомещения"])
print("ВидПомещения удалён (столбец)")

# b. Фильтрация статусов.
print("\nСтатусы до фильтрации:")
display(dataF_new['СледующийСтатус'].value_counts(dropna=False))

# Удаляем строчки с "В резерве" и пустые значения (NaN).
dataF_new = dataF_new[dataF_new['СледующийСтатус'].notna()]
dataF_new = dataF_new[dataF_new['СледующийСтатус'] != 'В резерве']

# "Продана" теперь будет - 1, а "Свободна" будет - 0.
dataF_new['СледующийСтатус'] = dataF_new['СледующийСтатус'].map({'Продана': 1, 'Свободна': 0})
print("\nСтатусы после фильтратии и преобразования (Продана - 1, Свободна - 0):")
display(dataF_new['СледующийСтатус'].value_counts())

# с. Удаление столбца УИД_Брони.
if 'УИД_Брони' in dataF_new.columns:
  dataF_new = dataF_new.drop(columns=['УИД_Брони'])
  print("\n'УИД_Брони удалён.")

# Результаты.
print("Строки и столбцы в отфильтрованном датасете:")
print(list(dataF_new.shape))

with pd.option_context('display.max_columns', None):
  print("\nПервые 10 строк отфильтрованных данных:")
  display(dataF_new.head(10))

Количество строк до фильтрации по жилым помещениям: 5519
Количество строк после фильтрации по жилым помещениям: 4105
ВидПомещения удалён (столбец)

Статусы до фильтрации:


СледующийСтатус
Свободна     2804
Продана      1140
NaN            85
В резерве      76
Name: count, dtype: int64


Статусы после фильтратии и преобразования (Продана - 1, Свободна - 0):


СледующийСтатус
0    2804
1    1140
Name: count, dtype: int64


'УИД_Брони удалён.
Строки и столбцы в отфильтрованном датасете:
[3944, 19]

Первые 10 строк отфильтрованных данных:


ДатаБрони ВремяБрони ИсточникБрони ВременнаяБронь  СледующийСтатус  \
0   13.08.2021    1:00:01        ручная             Да                0   
1   16.08.2021    4:12:46        ручная             Да                0   
2   16.08.2021    5:37:12        ручная             Да                0   
3   09.08.2021    4:17:35        ручная             Да                0   
4   16.08.2021    4:13:35        ручная             Да                0   
5   16.08.2021    9:30:15            МП            Нет                0   
6   18.08.2021    9:20:15            МП            Нет                0   
7   19.08.2021   10:20:16            МП            Нет                0   
9   21.08.2021    1:30:21            МП            Нет                0   
10  23.08.2021    9:10:15            МП            Нет                0   

               Город   Тип ПродаваемаяПлощадь  Этаж СтоимостьНаДатуБрони  \
0          Ярославль  2,5к                 72   3.0              4296100   
1          Ярославль  3,5к               79,8   2.0              5279600   
2          Ярославль  2,5к               65,4  18.0              4201400   
3             Тюмень  2,5к               61,6  12.0              5414500   
4          Ярославль  2,5к               65,4  18.0              4201400   
5   Набережные Челны  1,5к                 47   3.0              3494600   
6             Тюмень  3,5к               97,4   8.0              7535000   
7             Тюмень    2к               55,4   4.0              4954700   
9   Набережные Челны    2к               52,1   4.0              4030800   
10            Ижевск  3,5к                 87   2.0              5257900   

                 ТипСтоимости          ВариантОплаты ВариантОплатыДоп  \
0   Стоимость при 100% оплате  Единовременная оплата              NaN   
1   Стоимость при 100% оплате  Единовременная оплата              NaN   
2   Стоимость при 100% оплате  Единовременная оплата              NaN   
3   Стоимость при 100% оплате  Единовременная оплата              NaN   
4   Стоимость при 100% оплате  Единовременная оплата              NaN   
5   Стоимость при 100% оплате  Единовременная оплата          Ипотека   
6   Стоимость при 100% оплате  Единовременная оплата          Ипотека   
7   Стоимость при 100% оплате  Единовременная оплата          Ипотека   
9   Стоимость при 100% оплате  Единовременная оплата          Ипотека   
10  Стоимость при 100% оплате  Единовременная оплата              NaN   

   СкидкаНаКвартиру ФактическаяСтоимостьПомещения СделкаАН  \
0               NaN                       4296100      Нет   
1               NaN                       5279600       Да   
2               NaN                       4201400      Нет   
3               NaN                       5414500      Нет   
4               NaN                       4201400      Нет   
5               NaN                       3494600      Нет   
6               NaN                       7535000       Да   
7               NaN                       4954700      Нет   
9               NaN                       4030800      Нет   
10              NaN                       5257900       Да   

   ИнвестиционныйПродукт Привилегия Статус лида (из CRM)  
0                    Нет        Нет                    S  
1                    Нет        Нет                    F  
2                    Нет        Нет                    S  
3                    Нет        Нет                    S  
4                    Нет        Нет                    P  
5                    Нет        Нет                    P  
6                    Нет        Нет                    S  
7                    Нет        Нет                    P  
9                    Нет        Нет                    S  
10                   Нет        Нет                    P

Оставили только жилые помещения, удалили столбец "ВидПомещения" по условию задачи, так как он больше нам ненужен из-за отсутсвия других вариантов, избыточный столбец. Удалили лишние статусы (В резерве и NaN, как просили в задаче), преобразовали текстовые статусы в числовые. Продана - 1, а свободна - 0. Удалили столбец "УИД_Брони" - уникальный индентификатор брони, не нужная информация по условию задания. Естественно вывели данные до манипуляций и после, чтобы удостовериться в правильности.

In [ ]:
# Пункт 3 - проверка типов данных и преобразование к числовому типу.
# a. Проверка числовых полей.
print("\na. Проверка числовых полей:")
numeric_columns = ['ПродаваемаяПлощадь', 'Этаж', 'СтоимостьНаДатуБрони', 'СкидкаНаКвартиру', 'ФактическаяСтоимостьПомещения']

for columns in numeric_columns:
  if columns in dataF_new.columns:
    print(f"{columns}: {dataF_new[columns].dtype}")

# Показывает object поля, вместо числового формата, исправляем.
for columns in numeric_columns:
  if columns in dataF_new.columns and dataF_new[columns].dtype == 'object':
    dataF_new[columns] = dataF_new[columns].astype(str).str.replace(' ', '', regex=False)
    dataF_new[columns] = dataF_new[columns].str.replace(',', '.', regex=False)
    dataF_new[columns] = pd.to_numeric(dataF_new[columns], errors='coerce')

print("\nПроверка типов данных после преобразования:")
for columns in numeric_columns:
    if columns in dataF_new.columns:
        print(f"{columns}: {dataF_new[columns].dtype}")

# b. Преобразование бинарных признаков.
print("\nb. Преобразование бинарных признаков:")
if 'ИсточникБрони' in dataF_new.columns:
    dataF_new['ИсточникБрони'] = dataF_new['ИсточникБрони'].map({'ручная': 0, 'МП': 1})
    print("ИсточникБрони преобразован: ручная - 0, МП (Мобильное приложение) - 1")
if 'ВременнаяБронь' in dataF_new.columns:
    dataF_new['ВременнаяБронь'] = dataF_new['ВременнаяБронь'].map({'Нет': 0, 'Да': 1})
    print("ВременнаяБронь преобразован: Нет - 0, Да - 1")
if 'СделкаАН' in dataF_new.columns:
    dataF_new['СделкаАН'] = dataF_new['СделкаАН'].map({'Нет': 0, 'Да': 1})
    print("СделкаАН преобразован: Нет - 0, Да - 1")
if 'ИнвестиционныйПродукт' in dataF_new.columns:
    dataF_new['ИнвестиционныйПродукт'] = dataF_new['ИнвестиционныйПродукт'].map({'Нет': 0, 'Да': 1})
    print("ИнвестиционныйПродукт преобразован: Нет - 0, Да - 1")
if 'Привилегия' in dataF_new.columns:
    dataF_new['Привилегия'] = dataF_new['Привилегия'].map({'Нет': 0, 'Да': 1})
    print("Привилегия преобразован: Нет - 0, Да - 1")
if 'ВариантОплаты' in dataF_new.columns:
    dataF_new['ВариантОплаты'] = dataF_new['ВариантОплаты'].map({'Единовременная оплата': 0, 'Оплата в рассрочку': 1})
    print("ВариантОплаты преобразован: Единовременная оплата - 0, Оплата в рассрочку - 1")
if 'ТипСтоимости' in dataF_new.columns:
    dataF_new['ТипСтоимости'] = dataF_new['ТипСтоимости'].map({'Стоимость при 100% оплате': 0, 'Стоимость в рассрочку': 1})
    print("ТипСтоимости преобразован: Стоимость при 100% оплате - 0, Стоимость в рассрочку - 1")

# c. one-hot кодирование категориальных, небинарных признаков.
print("\nc. Кодирование категориальных, небинарных признаков:")
categorical_columns = ['Город', 'ВариантОплатыДоп', 'Статус лида (из CRM)']
available_categorical = [col for col in categorical_columns if col in dataF_new.columns]
if available_categorical:
    print(f"Категориальные признаки для one-hot кодирования: {available_categorical}")
    for col in available_categorical:
        print(f"\n{col}:")
        print(f"Тип данных: {dataF_new[col].dtype}")

        dummies = pd.get_dummies(dataF_new[col], prefix=col, dtype=int)
        # Добавляем новые столбцы к данным.
        dataF_new = pd.concat([dataF_new, dummies], axis=1)
        # Удаляем исходный столбец.
        dataF_new = dataF_new.drop(columns=[col])

        print(f"Создано {len(dummies.columns)} бинарных столбцов")
else:
    print("Категориальных признаков для one-hot кодирования не найдено")

# d. Преобразование поля "Тип" (убираем лишнее).
print("\nd. Преобразование поля 'Тип' (убираем лишнее):")
if 'Тип' in dataF_new.columns:
    def convert_room_type(room_str):
        if pd.isna(room_str):
            return np.nan
        room_str = str(room_str).strip()
        if room_str == 'с':
            return 1.0
        if room_str.endswith('к'):
            return float(room_str.replace('к', '').replace(',', '.'))
        try:
            return float(room_str)
        except:
            return np.nan

    dataF_new['Тип'] = dataF_new['Тип'].apply(convert_room_type)
    print('Поле "Тип" преобразовано')
else:
    print("Поле 'Тип' не найдено")

with pd.option_context('display.max_columns', None):
    print("\nПервые 10 строк после преобразования типов:")
    display(dataF_new.head(10))


a. Проверка числовых полей:
ПродаваемаяПлощадь: object
Этаж: float64
СтоимостьНаДатуБрони: object
СкидкаНаКвартиру: object
ФактическаяСтоимостьПомещения: object

Проверка типов данных после преобразования:
ПродаваемаяПлощадь: float64
Этаж: float64
СтоимостьНаДатуБрони: float64
СкидкаНаКвартиру: float64
ФактическаяСтоимостьПомещения: float64

b. Преобразование бинарных признаков:
ИсточникБрони преобразован: ручная - 0, МП (Мобильное приложение) - 1
ВременнаяБронь преобразован: Нет - 0, Да - 1
СделкаАН преобразован: Нет - 0, Да - 1
ИнвестиционныйПродукт преобразован: Нет - 0, Да - 1
Привилегия преобразован: Нет - 0, Да - 1
ВариантОплаты преобразован: Единовременная оплата - 0, Оплата в рассрочку - 1
ТипСтоимости преобразован: Стоимость при 100% оплате - 0, Стоимость в рассрочку - 1

c. Кодирование категориальных, небинарных признаков:
Категориальные признаки для one-hot кодирования: ['Город', 'ВариантОплатыДоп', 'Статус лида (из CRM)']

Город:
Тип данных: object
Создано 9 бинарных столб

ДатаБрони ВремяБрони  ИсточникБрони  ВременнаяБронь  СледующийСтатус  \
0   13.08.2021    1:00:01              0               1                0   
1   16.08.2021    4:12:46              0               1                0   
2   16.08.2021    5:37:12              0               1                0   
3   09.08.2021    4:17:35              0               1                0   
4   16.08.2021    4:13:35              0               1                0   
5   16.08.2021    9:30:15              1               0                0   
6   18.08.2021    9:20:15              1               0                0   
7   19.08.2021   10:20:16              1               0                0   
9   21.08.2021    1:30:21              1               0                0   
10  23.08.2021    9:10:15              1               0                0   

    Тип  ПродаваемаяПлощадь  Этаж  СтоимостьНаДатуБрони  ТипСтоимости  \
0   2.5                72.0   3.0             4296100.0             0   
1   3.5                79.8   2.0             5279600.0             0   
2   2.5                65.4  18.0             4201400.0             0   
3   2.5                61.6  12.0             5414500.0             0   
4   2.5                65.4  18.0             4201400.0             0   
5   1.5                47.0   3.0             3494600.0             0   
6   3.5                97.4   8.0             7535000.0             0   
7   2.0                55.4   4.0             4954700.0             0   
9   2.0                52.1   4.0             4030800.0             0   
10  3.5                87.0   2.0             5257900.0             0   

    ВариантОплаты  СкидкаНаКвартиру  ФактическаяСтоимостьПомещения  СделкаАН  \
0             0.0               NaN                      4296100.0         0   
1             0.0               NaN                      5279600.0         1   
2             0.0               NaN                      4201400.0         0   
3             0.0               NaN                      5414500.0         0   
4             0.0               NaN                      4201400.0         0   
5             0.0               NaN                      3494600.0         0   
6             0.0               NaN                      7535000.0         1   
7             0.0               NaN                      4954700.0         0   
9             0.0               NaN                      4030800.0         0   
10            0.0               NaN                      5257900.0         1   

    ИнвестиционныйПродукт  Привилегия  Город_Ижевск  Город_Набережные Челны  \
0                       0           0             0                       0   
1                       0           0             0                       0   
2                       0           0             0                       0   
3                       0           0             0                       0   
4                       0           0             0                       0   
5                       0           0             0                       1   
6                       0           0             0                       0   
7                       0           0             0                       0   
9                       0           0             0                       1   
10                      0           0             1                       0   

    Город_Пермь  Город_Сочи  Город_Тверь  Город_Тюмень  Город_Уфа  \
0             0           0            0             0          0   
1             0           0            0             0          0   
2             0           0            0             0          0   
3             0           0            0             1          0   
4             0           0            0             0          0   
5             0           0            0             0          0   
6             0           0            0             1          0   
7             0           0            0          

Проверили типы числовых данных, убедились, что поля имеют тип object, вместо числового и поэтому исправили этот недочёт путём удаления лишних элементов из этих столбцов (запятые, пробелы), далее для бинарных признаков сделали кодирование формата 0, 1, используя метод map(). Также сделали one-hot кодирование категориальных признаков, каждое уникальное значение стало отдельным столбцом. Преобразовали поле "Тип", убрав лишние элементы (если это не число). Естественно вывели новые данные, чтобы убедиться в правильности и корректности выполнения.

In [ ]:
# Пункт 4 - Проверка пропущенных данных.
print("\nПроверка наличия пропущенных данных:")
missing_data = dataF_new.isnull().sum()
missing_with_data = missing_data[missing_data > 0]

if len(missing_with_data) > 0:
    print("Столбцы с пропущенными значениями:")
    for column, count in missing_with_data.items():
        print(f"  {column}: {count} пропусков")
else:
    print(" Пропущенных значений не обнаружено")

# a. Обработка поля "СкидкаНаКвартиру".
print("\na. Обработка поля 'СкидкаНаКвартиру':")
if 'СкидкаНаКвартиру' in dataF_new.columns:
    before = dataF_new['СкидкаНаКвартиру'].isnull().sum()
    dataF_new['СкидкаНаКвартиру'] = dataF_new['СкидкаНаКвартиру'].fillna(0)
    after = dataF_new['СкидкаНаКвартиру'].isnull().sum()
    print(f" Заменено {before - after} пропусков на 0 (нет скидки)")

# b. Обработка полей "Тип" и "ПродаваемаяПлощадь".
print("\nb. Обработка полей 'Тип' и 'ПродаваемаяПлощадь':")
for column in ['Тип', 'ПродаваемаяПлощадь']:
    if column in dataF_new.columns:
        median_value = dataF_new[column].median()
        before = dataF_new[column].isnull().sum()
        dataF_new[column] = dataF_new[column].fillna(median_value)
        after = dataF_new[column].isnull().sum()
        print(f"  {column}: заменено {before - after} пропусков на медиану {median_value:.2f}")

# c. Обработка поля "ВариантОплатыДоп".
print("\nc. Обработка поля 'ВариантОплатыДоп':")
# Ищем one-hot столбцы, созданные в пункте 3.
oplata_dop_columns = [column for column in dataF_new.columns if column.startswith('ВариантОплатыДоп_')]

if oplata_dop_columns:
    print(f"  Поле преобразовано в one-hot: {len(oplata_dop_columns)} бинарных столбцов")
    for column in oplata_dop_columns:
        missing = dataF_new[column].isnull().sum()
        if missing > 0:
            dataF_new[column] = dataF_new[column].fillna(0)
            print(f"В '{column}' заменено {missing} пропусков на 0")
else:
    if 'ВариантОплатыДоп' in dataF_new.columns:
        print("  Столбец 'ВариантОплатыДоп' найден в исходном виде")
        # Проверяем, есть ли столбец "ВариантОплаты" для заполнения пропусков.
        if 'ВариантОплаты' in dataF_new.columns:
            before = dataF_new['ВариантОплатыДоп'].isnull().sum()
            dataF_new['ВариантОплатыДоп'] = dataF_new['ВариантОплатыДоп'].fillna(dataF_new['ВариантОплаты'])
            after = dataF_new['ВариантОплатыДоп'].isnull().sum()
            print(f"Заменено {before - after} пропусков значением из 'ВариантОплаты'")
        else:
            print(" Столбец 'ВариантОплаты' не найден, удаляем 'ВариантОплатыДоп'")
            dataF_new = dataF_new.drop(columns=['ВариантОплатыДоп'])
    else:
        print(" Поле 'ВариантОплатыДоп' не найдено ни в каком виде")

# d. Обработка остальных пропусков.
print("\nd. Обработка остальных пропусков:")
remaining_missing = dataF_new.isnull().sum()
remaining_with_missing = remaining_missing[remaining_missing > 0]

if len(remaining_with_missing) > 0:
    print("Остались пропуски в столбцах:")
    for column, count in remaining_with_missing.items():
        print(f"{column}: {count} пропусков")
    rows_with_missing = dataF_new.isnull().any(axis=1).sum()
    print(f"\n  Строк с пропусками: {rows_with_missing}")
    # Если строк с пропусками менее 10% от общего числа, удаляем их.
    if rows_with_missing < len(dataF_new) * 0.10:
        print(" Пропусков мало (<10%), удаляем строки с пропусками")
        rows_before = len(dataF_new)
        dataF_new = dataF_new.dropna()
        rows_after = len(dataF_new)
        print(f"Удалено строк: {rows_before - rows_after}")
        print(f"Осталось строк: {rows_after}")
    else:
        print("Пропусков много (>10%), оставляем как есть")
else:
    print("Пропусков больше нет")

print(f"\nИтоговый размер данных: {dataF_new.shape}")

with pd.option_context('display.max_columns', None):
    print("\nПервые 10 строк после обработки:")
    display(dataF_new.head(10))


Проверка наличия пропущенных данных:
Столбцы с пропущенными значениями:
  Тип: 44 пропусков
  ПродаваемаяПлощадь: 44 пропусков
  ВариантОплаты: 3 пропусков
  СкидкаНаКвартиру: 3246 пропусков

a. Обработка поля 'СкидкаНаКвартиру':
 Заменено 3246 пропусков на 0 (нет скидки)

b. Обработка полей 'Тип' и 'ПродаваемаяПлощадь':
  Тип: заменено 44 пропусков на медиану 2.00
  ПродаваемаяПлощадь: заменено 44 пропусков на медиану 60.50

c. Обработка поля 'ВариантОплатыДоп':
  Поле преобразовано в one-hot: 2 бинарных столбцов

d. Обработка остальных пропусков:
Остались пропуски в столбцах:
ВариантОплаты: 3 пропусков

  Строк с пропусками: 3
 Пропусков мало (<10%), удаляем строки с пропусками
Удалено строк: 3
Осталось строк: 3941

Итоговый размер данных: (3941, 30)

Первые 10 строк после обработки:


ДатаБрони ВремяБрони  ИсточникБрони  ВременнаяБронь  СледующийСтатус  \
0   13.08.2021    1:00:01              0               1                0   
1   16.08.2021    4:12:46              0               1                0   
2   16.08.2021    5:37:12              0               1                0   
3   09.08.2021    4:17:35              0               1                0   
4   16.08.2021    4:13:35              0               1                0   
5   16.08.2021    9:30:15              1               0                0   
6   18.08.2021    9:20:15              1               0                0   
7   19.08.2021   10:20:16              1               0                0   
9   21.08.2021    1:30:21              1               0                0   
10  23.08.2021    9:10:15              1               0                0   

    Тип  ПродаваемаяПлощадь  Этаж  СтоимостьНаДатуБрони  ТипСтоимости  \
0   2.5                72.0   3.0             4296100.0             0   
1   3.5                79.8   2.0             5279600.0             0   
2   2.5                65.4  18.0             4201400.0             0   
3   2.5                61.6  12.0             5414500.0             0   
4   2.5                65.4  18.0             4201400.0             0   
5   1.5                47.0   3.0             3494600.0             0   
6   3.5                97.4   8.0             7535000.0             0   
7   2.0                55.4   4.0             4954700.0             0   
9   2.0                52.1   4.0             4030800.0             0   
10  3.5                87.0   2.0             5257900.0             0   

    ВариантОплаты  СкидкаНаКвартиру  ФактическаяСтоимостьПомещения  СделкаАН  \
0             0.0               0.0                      4296100.0         0   
1             0.0               0.0                      5279600.0         1   
2             0.0               0.0                      4201400.0         0   
3             0.0               0.0                      5414500.0         0   
4             0.0               0.0                      4201400.0         0   
5             0.0               0.0                      3494600.0         0   
6             0.0               0.0                      7535000.0         1   
7             0.0               0.0                      4954700.0         0   
9             0.0               0.0                      4030800.0         0   
10            0.0               0.0                      5257900.0         1   

    ИнвестиционныйПродукт  Привилегия  Город_Ижевск  Город_Набережные Челны  \
0                       0           0             0                       0   
1                       0           0             0                       0   
2                       0           0             0                       0   
3                       0           0             0                       0   
4                       0           0             0                       0   
5                       0           0             0                       1   
6                       0           0             0                       0   
7                       0           0             0                       0   
9                       0           0             0                       1   
10                      0           0             1                       0   

    Город_Пермь  Город_Сочи  Город_Тверь  Город_Тюмень  Город_Уфа  \
0             0           0            0             0          0   
1             0           0            0             0          0   
2             0           0            0             0          0   
3             0           0            0             1          0   
4             0           0            0             0          0   
5             0           0            0             0          0   
6             0           0            0             1          0   
7             0           0            0          

Заменили все пропуски в "СкидкаНаКвартиру" на нули (скидки нет). Тип и ПродаваемаяПлощадь - пропуски заменили на медиану. Также, если строк с пропусками меньше 10%, то удаляем эти строки, а если больше, то оставляем. Конечно же выводим данные для просмотра и проверки.

In [ ]:
# Пункт 5 - дополнение данных.
# a. Цена за квадратный метр.
if 'ФактическаяСтоимостьПомещения' in dataF_new.columns and 'ПродаваемаяПлощадь' in dataF_new.columns:
    dataF_new['Цена за квадратный метр'] = dataF_new['ФактическаяСтоимостьПомещения'] / dataF_new['ПродаваемаяПлощадь']
    print("Новый признак 'Цена за квадратный метр'")
else:
    print("Не удалось добавить признак 'Цена за квадратный метр', так как отсутствуют необходимые столбцы.")

# b. Скидка в процентах.
if 'СкидкаНаКвартиру' in dataF_new.columns and 'СтоимостьНаДатуБрони' in dataF_new.columns:
    dataF_new['Скидка в процентах'] = dataF_new.apply(
        lambda row: (row['СкидкаНаКвартиру'] / row['СтоимостьНаДатуБрони'] * 100) if row['СтоимостьНаДатуБрони'] != 0 else 0,
        axis=1
    )
    print("Новый признак 'Скидка в процентах'")
else:
    print("Не удалось добавить признак 'Скидка в процентах', так как отсутствуют необходимые столбцы")

# Проверим добавленные признаки.
print("\nСтатистика новых признаков:")
if 'Цена за квадратный метр' in dataF_new.columns:
    print(f"Цена за квадратный метр: от {dataF_new['Цена за квадратный метр'].min():.2f} до {dataF_new['Цена за квадратный метр'].max():.2f}")

if 'Скидка в процентах' in dataF_new.columns:
    print(f"Скидка в процентах: от {dataF_new['Скидка в процентах'].min():.2f} до {dataF_new['Скидка в процентах'].max():.2f}")

with pd.option_context('display.max_columns', None):
    print("\nПервые 10 строк с новыми признаками:")
    display_columns = ['ФактическаяСтоимостьПомещения', 'ПродаваемаяПлощадь', 'Цена за квадратный метр',
                   'СкидкаНаКвартиру', 'СтоимостьНаДатуБрони', 'Скидка в процентах']
    available_columns = [col for col in display_columns if col in dataF_new.columns]
    display(dataF_new[available_columns].head(10))

Новый признак 'Цена за квадратный метр'
Новый признак 'Скидка в процентах'

Статистика новых признаков:
Цена за квадратный метр: от 44467.05 до 463553.72
Скидка в процентах: от -8.00 до 13.44

Первые 10 строк с новыми признаками:


ФактическаяСтоимостьПомещения  ПродаваемаяПлощадь  \
0                       4296100.0                72.0   
1                       5279600.0                79.8   
2                       4201400.0                65.4   
3                       5414500.0                61.6   
4                       4201400.0                65.4   
5                       3494600.0                47.0   
6                       7535000.0                97.4   
7                       4954700.0                55.4   
9                       4030800.0                52.1   
10                      5257900.0                87.0   

    Цена за квадратный метр  СкидкаНаКвартиру  СтоимостьНаДатуБрони  \
0              59668.055556               0.0             4296100.0   
1              66160.401003               0.0             5279600.0   
2              64241.590214               0.0             4201400.0   
3              87897.727273               0.0             5414500.0   
4              64241.590214               0.0             4201400.0   
5              74353.191489               0.0             3494600.0   
6              77361.396304               0.0             7535000.0   
7              89435.018051               0.0             4954700.0   
9              77366.602687               0.0             4030800.0   
10             60435.632184               0.0             5257900.0   

    Скидка в процентах  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
5                  0.0  
6                  0.0  
7                  0.0  
9                  0.0  
10                 0.0

Создаём новые признаки "цена за квадратный метр" и "скидка в процентах". Для вычисления цены за квадратный метр, стоимость помещения на продаваемую площадь. Для вычисления скидки в процентах, мы делим скидку на квартиру на СтоимостьНаДатуБрони, а потом умножаем на 100, чтобы перевести в проценты, и конечно проверяем деление на ноль. Если же стоимость 0, то просто возвращаем 0%.

In [ ]:
# Пункт 6 - нормализация данных.
# Сохраняем исходные данные.
dataF_before_norm = dataF_new.copy()

# Находим числовые признаки.
numerical_columns = []
for column in dataF_new.columns:
    if pd.api.types.is_numeric_dtype(dataF_new[column]) and column != 'СледующийСтатус':
        numerical_columns.append(column)

print(f"Числовых признаков для нормализации: {len(numerical_columns)}")

# Обычные признаки и 'СкидкаНаКвартиру'.
columns_standard = [col for col in numerical_columns if col != 'СкидкаНаКвартиру']
discount_column = ['СкидкаНаКвартиру'] if 'СкидкаНаКвартиру' in numerical_columns else []

print(f"Обычных признаков: {len(columns_standard)}")
if discount_column:
    print(f"Особый признак: 'СкидкаНаКвартиру'")

# 1. Нормализация обычных признаков в [0, 1].
if columns_standard:
    scaler_standard = MinMaxScaler(feature_range=(0, 1))
    dataF_new[columns_standard] = scaler_standard.fit_transform(dataF_new[columns_standard])
    print("Обычные признаки нормализованы в [0, 1]")

# 2. 'СкидкаНаКвартиру' в [-0.5, 0.5].
if discount_column:
    scaler_discount = MinMaxScaler(feature_range=(-0.5, 0.5))
    dataF_new[discount_column] = scaler_discount.fit_transform(dataF_new[discount_column])
    print("'СкидкаНаКвартиру' нормализована в [-0.5, 0.5]")

# Проверяем результаты.
print("\nПроверка нормализации:")
if columns_standard[:5]:
    for column in columns_standard[:5]:
        print(f"{column}: было [{dataF_before_norm[column].min():.2f}, {dataF_before_norm[column].max():.2f}], стало [{dataF_new[column].min():.2f}, {dataF_new[column].max():.2f}]")
if discount_column:
    column = discount_column[0]
    print(f"\n{column}: было [{dataF_before_norm[column].min():.2f}, {dataF_before_norm[column].max():.2f}], стало [{dataF_new[column].min():.2f}, {dataF_new[column].max():.2f}]")
print(f"\nИтоговый размер: {dataF_new.shape}")

Числовых признаков для нормализации: 29
Обычных признаков: 28
Особый признак: 'СкидкаНаКвартиру'
Обычные признаки нормализованы в [0, 1]
'СкидкаНаКвартиру' нормализована в [-0.5, 0.5]

Проверка нормализации:
ИсточникБрони: было [0.00, 1.00], стало [0.00, 1.00]
ВременнаяБронь: было [0.00, 1.00], стало [0.00, 1.00]
Тип: было [1.00, 4.50], стало [0.00, 1.00]
ПродаваемаяПлощадь: было [27.30, 139.65], стало [0.00, 1.00]
Этаж: было [1.00, 25.00], стало [0.00, 1.00]

СкидкаНаКвартиру: было [-838881.00, 1334700.00], стало [-0.50, 0.50]

Итоговый размер: (3941, 32)


-0.5 - максимальная наценка, а 0.5 - максимальная скидка, для скидки на квартиру больше подходит этот диаапазон, так как при 0 и 1 нельзя будет нормально отличить скидку и наценку, это очень удобный диапазон, при рассмотрении числа, например в -500000 - -0.5 мы видим, что наценка огромная.

In [ ]:
# Пункт 7 - проверка сбалансированности данных.
if 'СледующийСтатус' in dataF_new.columns:
    # Считаем количество записей.
    balance_counts = dataF_new['СледующийСтатус'].value_counts()
    print(f"Свободна - 0: {balance_counts.get(0, 0)} записей")
    print(f"Продана - 1: {balance_counts.get(1, 0)} записей")
    # Вычисляем проценты.
    total = len(dataF_new)
    percent_0 = (balance_counts.get(0, 0) / total) * 100
    percent_1 = (balance_counts.get(1, 0) / total) * 100

    print(f"\nПроцентное соотношение:")
    print(f"Свободна - 0: {percent_0:.2f}%")
    print(f"Продана - 1: {percent_1:.2f}%")

    # Проверяем сбалансированность.
    print(f"\nПроверяем сбалансированность:")
    if min(percent_0, percent_1) >= 40:
        print("Датасет сбалансирован!")
        print(f"Меньшее значение из классов: {min(percent_0, percent_1):.2f}%")
    elif min(percent_0, percent_1) >= 30:
        print("Датасет немного несбалансирован!")
        print(f"Меньшее значение из классовс: {min(percent_0, percent_1):.2f}%")
    else:
        print("Датасет сильно несбалансирован!")
        print(f"Меньшее значение из классов: {min(percent_0, percent_1):.2f}%")
else:
    print("Целевой признак 'СледующийСтатус' не найден")

Свободна - 0: 2804 записей
Продана - 1: 1137 записей

Процентное соотношение:
Свободна - 0: 71.15%
Продана - 1: 28.85%

Проверяем сбалансированность:
Датасет сильно несбалансирован!
Меньшее значение из классов: 28.85%


Берём целевой признак, считаем количество записей у его классов и дальше смотрим по процентам. Если значение меньшего класса больше 40%, то значит, что датасет вполне сбалансирован, если больше 30, но меньше 40, то он немного несбалансирован, но если же значение меньше 30%, то значит, что он серьёзно не сбалансирован. В нашем случае как раз такой вариант - 28.85%

In [ ]:
# Пункт 8 - формирование признаков.
# 1. Целевой признак.
target_column = 'СледующийСтатус'
if target_column in dataF_new.columns:
    y = dataF_new[target_column]
    print(f"Целевой признак: '{target_column}'")
    print(f"Размер y: {y.shape}")
else:
    print(f"Целевой признак '{target_column}' не найден")
    y = None

# 2. Факторные признаки.
feature_columns = []
for column in dataF_new.columns:
    if column == target_column:
        continue
    if pd.api.types.is_numeric_dtype(dataF_new[column]):
        feature_columns.append(column)
if feature_columns:
    x = dataF_new[feature_columns]
    print(f"\nФакторные признаки: {len(feature_columns)} столбцов")
    print(f"Размер x: {x.shape}")
else:
  print("Не найдено факторных признаков")
  x = None

Целевой признак: 'СледующийСтатус'
Размер y: (3941,)

Факторные признаки: 29 столбцов
Размер x: (3941, 29)


Сформировали целевой признак - y и факторный признак - x.

In [ ]:
# Пункт 9 - разбиение на обучающую и тестовую выборки.
if x is not None and y is not None:
    # Разбиваем в соотношении 80% обучающая на 20% тестовая
    x_train, x_test, y_train, y_test = train_test_split(
        x, y,
        test_size=0.2,
        random_state=5,
        stratify=y
    )

    print("Данные успешно разделены:")
    print(f"Обучающая выборка: {x_train.shape[0]} строк, {x_train.shape[1]} признаков")
    print(f"Тестовая выборка: {x_test.shape[0]} строк, {x_test.shape[1]} признаков")
    # Проверяем распределение классов.
    print("\nРаспределение классов в выборках:")
    # Обучающая.
    train_0 = (y_train == 0).sum()
    train_1 = (y_train == 1).sum()
    train_total = len(y_train)
    # Тестовая.
    test_0 = (y_test == 0).sum()
    test_1 = (y_test == 1).sum()
    test_total = len(y_test)

    print(f"Обучающая:")
    print(f"Класс 0: {train_0} ({train_0/train_total*100:.2f}%)")
    print(f"Класс 1: {train_1} ({train_1/train_total*100:.2f}%)")

    print(f"Тестовая:")
    print(f"Класс 0: {test_0} ({test_0/test_total*100:.2f}%)")
    print(f"Класс 1: {test_1} ({test_1/test_total*100:.2f}%)")
    original_0 = (y == 0).sum()
    original_1 = (y == 1).sum()
    original_total = len(y)

    print(f"\nИсходные данные:")
    print(f"Класс 0: {original_0} ({original_0/original_total*100:.2f}%)")
    print(f"Класс 1: {original_1} ({original_1/original_total*100:.2f}%)")

else:
    print("Не удалось разделить данные. Проверьте x и y.")

NameError: name 'x' is not defined

Разбиваем датасет на обучающую и тестовую выборку (пропорции 80 на 20), вписываем random_state=5, чтобы данные в случайном порядке брались, цифру 5 берём, чтобы сохранить наш "сид" (вариацию порядка), stratify=y нужен, чтобы наверняка соблюдались пропорции.

In [ ]:
# Пункт 10 - модель KNN.
if x_train is not None and y_train is not None:
    # 1. Создаём модель с параметрами по умолчанию.
    knn_model = KNeighborsClassifier()

    # 2. Обучаем модель на тренировочных данных.
    knn_model.fit(x_train, y_train)
    print("Модель KNN обучена успешно")
    knn_model.get_params()

    # 3. Получаем прогнозы.
    # Прогнозы на обучающей выборке.
    y_train_pred_knn = knn_model.predict(x_train)

    # Прогнозы на тестовой выборке.
    y_test_pred_knn = knn_model.predict(x_test)
    print(f"Прогнозы получены:")
    print(f"На обучающих данных: {len(y_train_pred_knn)} прогнозов")
    print(f"На тестовых данных: {len(y_test_pred_knn)} прогнозов")

    # 4. Распределение прогнозов
    print("\nРаспределение прогнозов KNN:")
    train_pred_0 = (y_train_pred_knn == 0).sum()
    train_pred_1 = (y_train_pred_knn == 1).sum()
    train_pred_total = len(y_train_pred_knn)

    test_pred_0 = (y_test_pred_knn == 0).sum()
    test_pred_1 = (y_test_pred_knn == 1).sum()
    test_pred_total = len(y_test_pred_knn)

    print(f"На обучающих данных:")
    print(f"Прогноз 0: {train_pred_0} ({train_pred_0/train_pred_total*100:.2f}%)")
    print(f"Прогноз 1: {train_pred_1} ({train_pred_1/train_pred_total*100:.2f}%)")

    print(f"На тестовых данных:")
    print(f"Прогноз 0: {test_pred_0} ({test_pred_0/test_pred_total*100:.2f}%)")
    print(f"Прогноз 1: {test_pred_1} ({test_pred_1/test_pred_total*100:.2f}%)")

    print(f"Реальное распределение в обучающих данных:")
    print(f"Класс 0: {train_0} ({train_0/train_total*100:.2f}%)")
    print(f"Класс 1: {train_1} ({train_1/train_total*100:.2f}%)")

else:
    print("Данные не подготовлены.")

Модель KNN обучена успешно
Прогнозы получены:
На обучающих данных: 3152 прогнозов
На тестовых данных: 789 прогнозов

Распределение прогнозов KNN:
На обучающих данных:
Прогноз 0: 2307 (73.19%)
Прогноз 1: 845 (26.81%)
На тестовых данных:
Прогноз 0: 573 (72.62%)
Прогноз 1: 216 (27.38%)
Реальное распределение в обучающих данных:
Класс 0: 2243 (71.16%)
Класс 1: 909 (28.84%)


Создаём алгоритм классификации - KNN. Строим и обучаем модель со стандартными параметрами.

In [ ]:
# Пункт 11 - модель Decision Tree.
if x_train is not None and y_train is not None:
    # 1. Создаём модель с параметрами по умолчанию.
    tree_model = DecisionTreeClassifier(random_state=5)
    # 2. Обучаем модель на тренировочных данных.
    tree_model.fit(x_train, y_train)
    print("Модель Decision Tree обучена успешно!")
    tree_model.get_params()
    # Прогнозы на обучающей выборке.
    y_train_pred_tree = tree_model.predict(x_train)

    # Прогнозы на тестовой выборке.
    y_test_pred_tree = tree_model.predict(x_test)
    print(f"Прогнозы получены:")
    print(f"На обучающих данных: {len(y_train_pred_tree)} прогнозов")
    print(f"На тестовых данных: {len(y_test_pred_tree)} прогнозов")

    # 3. Рспределение прогнозов.
    print("\nРаспределение прогнозов Decision Tree:")
    train_pred_0_tree = (y_train_pred_tree == 0).sum()
    train_pred_1_tree = (y_train_pred_tree == 1).sum()
    train_pred_total_tree = len(y_train_pred_tree)

    test_pred_0_tree = (y_test_pred_tree == 0).sum()
    test_pred_1_tree = (y_test_pred_tree == 1).sum()
    test_pred_total_tree = len(y_test_pred_tree)

    print(f"На обучающих данных:")
    print(f"Прогноз 0: {train_pred_0_tree} ({train_pred_0_tree/train_pred_total_tree*100:.2f}%)")
    print(f"Прогноз 1: {train_pred_1_tree} ({train_pred_1_tree/train_pred_total_tree*100:.2f}%)")

    print(f"На тестовых данных:")
    print(f"Прогноз 0: {test_pred_0_tree} ({test_pred_0_tree/test_pred_total_tree*100:.2f}%)")
    print(f"Прогноз 1: {test_pred_1_tree} ({test_pred_1_tree/test_pred_total_tree*100:.2f}%)")

    print(f"Реальное распределение в обучающих данных:")
    print(f"Класс 0: {train_0} ({train_0/train_total*100:.2f}%)")
    print(f"Класс 1: {train_1} ({train_1/train_total*100:.2f}%)")


Модель Decision Tree обучена успешно!
Прогнозы получены:
На обучающих данных: 3152 прогнозов
На тестовых данных: 789 прогнозов

Распределение прогнозов Decision Tree:
На обучающих данных:
Прогноз 0: 2252 (71.45%)
Прогноз 1: 900 (28.55%)
На тестовых данных:
Прогноз 0: 570 (72.24%)
Прогноз 1: 219 (27.76%)
Реальное распределение в обучающих данных:
Класс 0: 2243 (71.16%)
Класс 1: 909 (28.84%)


Создали и обучили модель "дерево решений". Структура алгоритма состоит в различных правилах "если есть это, то будет это" и т.д.

# Пункт 12 - векторы прогнозных значений целевой переменной на обучающей и на тестовой выборках для каждой из моделей.
Мы по сути их уже получали в прошлых пунктах, так что приступаем к пункту 13.

In [ ]:
# Пункт 13 - метрики качества.
# Проверяем прогнозы.
if ('y_train_pred_knn' in locals() and 'y_test_pred_knn' in locals() and
    'y_train_pred_tree' in locals() and 'y_test_pred_tree' in locals()):

    # Функция расчёта метрик.
    def calculate_metrics(y_true, y_pred):
        return {
            'F1': f1_score(y_true, y_pred),
            'Precision': precision_score(y_true, y_pred),
            'Recall': recall_score(y_true, y_pred)
        }
    # Рассчитываем метрики.
    knn_train = calculate_metrics(y_train, y_train_pred_knn)
    knn_test = calculate_metrics(y_test, y_test_pred_knn)
    tree_train = calculate_metrics(y_train, y_train_pred_tree)
    tree_test = calculate_metrics(y_test, y_test_pred_tree)

    # Выводим результаты.
    print("\n1. KNN модель:")
    print("Обучающая выборка:")
    print(f"F1: {knn_train['F1']:.4f}, Precision: {knn_train['Precision']:.4f}, Recall: {knn_train['Recall']:.4f}")
    print("Тестовая выборка:")
    print(f"F1: {knn_test['F1']:.4f}, Precision: {knn_test['Precision']:.4f}, Recall: {knn_test['Recall']:.4f}")

    print("\n2. Decision Tree модель:")
    print("Обучающая выборка:")
    print(f"F1: {tree_train['F1']:.4f}, Precision: {tree_train['Precision']:.4f}, Recall: {tree_train['Recall']:.4f}")
    print("Тестовая выборка:")
    print(f"F1: {tree_test['F1']:.4f}, Precision: {tree_test['Precision']:.4f}, Recall: {tree_test['Recall']:.4f}")

    # Сравнение моделей.
    print("\n3. Сравнение:")
    if knn_test['F1'] > tree_test['F1']:
        print(f"Лучшая модель: KNN (F1: {knn_test['F1']:.4f} > {tree_test['F1']:.4f})")
    else:
        print(f"Лучшая модель: Decision Tree (F1: {tree_test['F1']:.4f} > {knn_test['F1']:.4f})")


1. KNN модель:
Обучающая выборка:
F1: 0.8187, Precision: 0.8497, Recall: 0.7899
Тестовая выборка:
F1: 0.7207, Precision: 0.7407, Recall: 0.7018

2. Decision Tree модель:
Обучающая выборка:
F1: 0.9950, Precision: 1.0000, Recall: 0.9901
Тестовая выборка:
F1: 0.7517, Precision: 0.7671, Recall: 0.7368

3. Сравнение:
Лучшая модель: Decision Tree (F1: 0.7517 > 0.7207)


# Пункт 14 - вывод.
Дерево решений показало более высокую F1-меру, однако мы можем наблюдать переобучение, так как в обучающей выборке значительно более высокий показатель, чем в тестовой. KNN модель хоть и имеет более низкие показатели, но лучше справляется с новыми данными.